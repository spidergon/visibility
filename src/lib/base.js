import { useState, useEffect } from 'react'
import firebase from 'firebase/app'
import 'firebase/auth'
import 'firebase/firestore'
import 'firebase/storage'
import config from '../../firebase/config'
import { showSnack } from './state'

/*  ------------
    --- Init ---
    ------------  */

firebase.initializeApp(config)

export default firebase

export const auth = firebase.auth()

export const db = firebase.firestore()

export const storage = firebase.storage()

/*  ---------------
    --- Storage ---
    ---------------  */

/**
 * Upload a file to the storage.
 * @param {string} name - the name of the file.
 * @param {string} data - the data of the file.
 */
export async function uploadFile (name, data) {
  const uploadRes = await storage
    .ref('stores')
    .child(`${auth.currentUser.uid}/${Date.now()}_${name}`)
    .putString(data, 'data_url')
  return {
    name,
    src: await uploadRes.ref.getDownloadURL(),
    path: uploadRes.ref.location.path
  }
}

/**
 * Delete a file from the storage.
 * @param {string} path - the path of the file to delete.
 */
export function deleteFile (path) {
  return storage
    .ref()
    .child(path)
    .delete()
}

/*  -----------------
    --- FireStore ---
    -----------------  */

const storeCollectionName = 'stores'
const userCollectionName = 'users'

// Configure offline persistence
db.enablePersistence({ synchronizeTabs: true }).catch(err => {
  if (err.code === 'unimplemented') {
    // The current browser does not support all of the
    // features required to enable persistence.
    console.log("Data can't be persisted on your browser")
  }
})
// Subsequent queries will use persistence, if it was enabled successfully

/**
 * Check if a user exists.
 * @param {string} id the id of the user.
 */
export async function userExists (id) {
  const doc = await db
    .collection(userCollectionName)
    .doc(id)
    .get()
    .catch(() => {
      /* Silent is golden */
    })
  if (doc) return doc.exists
  return false
}

/**
 * Create a new user in the database.
 * @param {Object} data - the data of the new user.
 */
export async function addUser (data) {
  if (data && data.uid) {
    if (!(await userExists(data.uid))) {
      const uid = data.uid
      delete data.uid
      return db
        .collection(userCollectionName)
        .doc(uid)
        .set(data)
    }
  }
}

/**
 * Check if a store exists.
 * @param {string} id the id of the store.
 */
export async function storeExists (id) {
  const doc = await db
    .collection(storeCollectionName)
    .doc(id)
    .get()
    .catch(() => {
      /* Silent is golden */
    })
  if (doc) return doc.exists
  return false
}

/**
 * Create a new store in the database.
 * @param {Object} data - the data of the new store.
 */
export function addStore (data) {
  if (auth.currentUser && data && data.slug) {
    const slug = data.slug
    data.author = auth.currentUser.uid
    data.created = new Date()
    data.archived = false
    data.status = 'offline'
    data.fans = []
    data.errorMsg = ''
    delete data.slug
    return db
      .collection(storeCollectionName)
      .doc(slug)
      .set(data)
  }
}

/**
 * Update a store.
 * @param {string} id - the id of the store to be updated.
 * @param {Object} set - the object query to be set.
 * @param {string} successMsg  - message in case of success.
 * @param {string} failMsg - message in case of failure
 * @param {function} callback - the success callback.
 */
export function updateStore (id, set, successMsg, failMsg, callback) {
  if (id) {
    return db
      .collection(storeCollectionName)
      .doc(id)
      .set(set, { merge: true })
      .then(() => {
        showSnack(successMsg, 'success')
        if (typeof callback === 'function') callback()
      })
      .catch(err => {
        console.log(err)
        showSnack(failMsg, 'error')
      })
  }
}

/**
 * Delete a store (will be archived in the database).
 * @param {string} id - the id of the store to be archived.
 * @param {function} callback - the success callback.
 */
export function deleteStore (id, callback) {
  updateStore(
    id,
    { archived: true, status: 'offline' },
    'Votre vitrine a été supprimée avec succès.',
    'Suppression imposible.',
    callback
  )
}

/**
 * Publish a store.
 * @param {string} id - the id of the store to be published.
 * @param {function} callback - the success callback.
 */
export function publishStore (id, callback) {
  updateStore(
    id,
    { status: 'waiting' },
    'Votre vitrine est en attente de publication.',
    'Mise en ligne imposible.',
    callback
  )
}

/**
 * Add a fan to the store or remove it.
 * @param {*} storeId - the id of the store to be updated.
 * @param {*} userId - the user id to add or remove.
 * @param {*} callback - function to execute in case of success.
 * @param {*} isFan - whether the user is fan or not
 */
export function addRemoveStoreFav (storeId, userId, callback, isFan = false) {
  if (storeId && userId) {
    let fans = firebase.firestore.FieldValue.arrayUnion(userId)
    if (isFan) fans = firebase.firestore.FieldValue.arrayRemove(userId)
    return db
      .collection(storeCollectionName)
      .doc(storeId)
      .update({ fans })
      .then(() => {
        if (typeof callback === 'function') callback()
      })
      .catch(err => {
        console.log(err)
        showSnack("Une erreur interne s'est produite.", 'error')
      })
  }
}

/**
 * Hook that provides the store by its id.
 * @param {string} id - the id.
 * @returns {Object} { error, loading, store }.
 */
export function useStore (id) {
  const [error, setError] = useState(false)
  const [loading, setLoading] = useState(true)
  const [store, setStore] = useState(null)

  useEffect(() => {
    const unsubscribe = db
      .collection(storeCollectionName)
      .doc(id)
      .onSnapshot(
        doc => {
          if (doc && doc.exists) setStore({ id, ...doc.data() })
          setLoading(false)
        },
        err => setError(err)
      )
    return () => unsubscribe()
  }, [id])

  return { error, loading, store }
}

// /**
//  * Hook that provides the store by its id.
//  * @param {string} id - the id.
//  * @returns {Object} Object containing { loading, store }.
//  */
// export function useMyStore (id) {
//   const [loading, setLoading] = useState(null)
//   const [store, setStore] = useState(null)

//   useEffect(() => {
//     let cancelled = false
//     if (id) {
//       if (cancelled) return
//       setLoading(true)
//       db.collection(storeCollectionName)
//         .doc(id)
//         .get()
//         .then(doc => {
//           if (cancelled) return
//           if (doc && doc.exists) {
//             setStore({ id, ...doc.data() })
//           }
//           setLoading(false)
//         })
//         .catch(err => {
//           if (cancelled) return
//           setLoading(false)
//           console.log(err)
//           showSnack(`Vitrine introuvable : ${id}`, 'error')
//         })
//     }
//     return () => (cancelled = true) // Cleaning
//   }, [id])

//   return { loading, store }
// }

/** Get the first part of the query depending on user and fav */
function userFavQuery (user, fav) {
  const uid = user ? user.uid : null
  const collection = db.collection(storeCollectionName)
  if (fav) return collection.where('fans', 'array-contains', uid)
  return collection.where('author', '==', uid)
}

/**
 * Hook that provides the stores for a user.
 * @param {Object} user - the user object.
 * @returns {Object} { error, loading, stores }.
 */
export function useMyStores (user, fav = false) {
  const [error, setError] = useState(false)
  const [loading, setLoading] = useState(true)
  const [stores, setStores] = useState([])

  useEffect(() => {
    const unsubscribe = userFavQuery(user, fav)
      .where('archived', '==', false)
      .orderBy('created', 'desc')
      .limit(8)
      .onSnapshot(
        snapshot => {
          const myStores = []
          snapshot.forEach(doc => {
            const data = doc.data()
            data.loved = data.fans.includes(user ? user.uid : null)
            myStores.push({ id: doc.id, ...data })
          })
          setStores(myStores)
          setLoading(false)
        },
        err => setError(err)
      )
    return () => unsubscribe()
  }, [fav, user])

  return { error, loading, stores }
}

// /**
//  * Hook that provides the stores for a user.
//  * @param {Object} user - the user object.
//  * @returns {Object} Object containing { loading, stores }.
//  */
// export function useMyStores_ (user, fav = false) {
//   const [loading, setLoading] = useState(false)
//   const [stores, setStores] = useState([])

//   useEffect(() => {
//     // const cancel = myStores()
//     // return () => cancel() // Cleaning
//     let cancelled = false
//     if (!fav && user && user.uid) {
//       if (cancelled) return
//       setLoading(true)
//       db.collection(storeCollectionName)
//         .where('author', '==', user.uid)
//         .where('archived', '==', false)
//         .orderBy('created', 'desc')
//         .limit(8)
//         .get()
//         .then(snap => {
//           const newStores = []
//           snap.forEach(function (doc) {
//             newStores.push({ id: doc.id, ...doc.data() })
//           })
//           if (cancelled) return
//           setStores(newStores)
//           setLoading(false)
//         })
//         .catch(err => {
//           if (cancelled) return
//           setLoading(false)
//           console.log(err)
//           showSnack("Une erreur interne s'est produite.", 'error')
//         })
//     }
//     return () => (cancelled = true) // Cleaning
//   }, [fav, user])

//   return { loading, stores }
// }
