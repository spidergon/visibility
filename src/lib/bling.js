// $ of jQuery without the jQuery
// based on https://gist.github.com/paulirish/12fb951a8b893a454b32

const $ = document.querySelector.bind(document)
const $$ = document.querySelectorAll.bind(document)

Node.prototype.on = window.on = function (name, fn) {
  this.addEventListener(name, fn)
}

Object.setPrototypeOf(NodeList.prototype, Array.prototype)

NodeList.prototype.on = NodeList.prototype.addEventListener = function (
  name,
  fn
) {
  this.forEach(elem => {
    elem.on(name, fn)
  })
}

export { $, $$ }
