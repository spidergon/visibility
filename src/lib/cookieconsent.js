import 'cookieconsent'

function cookieIds () {
  return document.cookie.split(';').map(c => c.split('=', 1)[0].trim())
}

function enable () {
  window.isCookies = true
  // Google analytics
  window['ga-disable-UA-XXXXX-Y'] = false
}

function disable () {
  window.isCookies = false
  // Google analytics
  window['ga-disable-UA-XXXXX-Y'] = true
  // Delete cookies
  cookieIds().forEach(id => {
    if (id !== 'cookieconsent_status') {
      document.cookie = `${id}=; expires=Thu, 01 Jan 1970 00:00:00 GMT; path=/`
    }
  })
}

function statusChange (type, didConsent) {
  if (type === 'opt-in' && didConsent) enable()
  if (type === 'opt-out' && !didConsent) disable()
}

window.cookieconsent.initialise({
  palette: {
    popup: {
      background: '#000'
    },
    button: {
      background: 'transparent',
      text: '#fff',
      border: '#535353'
    }
  },
  position: 'bottom-left',
  type: 'opt-out',
  content: {
    message:
      "Nous utilisons des cookies afin d'améliorer votre expérience sur le site et ainsi vous proposer un contenu adapté à vos besoins.",
    link: 'En savoir plus',
    href: '/mentions-legales/#cookiepolicy',
    dismiss: 'Fermer',
    deny: 'Refuser',
    allow: 'Accepter'
  },
  onInitialise (status) {
    statusChange(this.options.type, this.hasConsented())
  },
  onStatusChange (status, chosenBefore) {
    statusChange(this.options.type, this.hasConsented())
  },
  onRevokeChoice () {
    const type = this.options.type
    if (type === 'opt-in') disable()
    if (type === 'opt-out') enable()
  }
})
