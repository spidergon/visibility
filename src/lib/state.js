import { createGlobalState } from 'react-hooks-global-state'

const defaultState = {
  snack: { msg: '', variant: '' },
  dash: { tabVal: 0 }
}

export const {
  GlobalStateProvider,
  setGlobalState,
  useGlobalState
} = createGlobalState(defaultState)

/**
 * Show a snackbar containing a message.
 * @param {string} msg - the message to display.
 * @param {string} [variant] - the variant of the message (default: 'info').
 */
export function showSnack (msg, variant = 'info') {
  setGlobalState('snack', { msg, variant })
}

/**
 * Set the current tab value in the dashboard.
 * @param {number} tabVal - the tab value.
 */
export function setTabVal (tabVal) {
  setGlobalState('dash', v => ({ ...v, tabVal }))
}
