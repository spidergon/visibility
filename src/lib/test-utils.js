import React from 'react'
import PropTypes from 'prop-types'
import { render } from '@testing-library/react'
import { ThemeProvider } from 'styled-components'
import { theme, Style } from '../components/Style'

const AllTheProviders = ({ children }) => (
  <ThemeProvider theme={theme}>
    <>
      <Style />
      {children}
    </>
  </ThemeProvider>
)

AllTheProviders.propTypes = {
  children: PropTypes.node.isRequired
}

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

// re-export everything
export * from '@testing-library/react'

// override render method
export { customRender as render }

// import casual from 'casual'
// import md5 from 'md5'

// casual.seed(777) // seed it so we get consistent results

// export const fakeUser = (isAnonymous = false, emailVerified = false) => ({
//   __typename: 'User',
//   uid: 'CUbE9RoUIYVOzGp04UPVzQO68Xr1',
//   isAnonymous,
//   emailVerified,
//   displayName: casual.name,
//   email: casual.email,
//   photoURL: `https://gravatar.com/avatar/${md5(casual.email)}?s=50`,
//   signOut () {}
// })

// export const fakeStore = (archived = false, status = 'online') => ({
//   __typename: 'Store',
//   id: 'ma-superbe-vitrine',
//   activity: 'Restaurant',
//   address: casual.address,
//   archived,
//   author: 'CUbE9RoUIYVOzGp04UPVzQO68Xr1',
//   company: casual.company_name,
//   coordinates: [casual.longitude, casual.latitude],
//   created: casual.date,
//   description: `{"blocks":[{"key":"6rjnn","text":"Ma Superbe Vitrine","type":"header-two","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"7ve8t","text":"Une vitrine Ã  visiter !","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}},{"key":"2fpj7","text":"By Chris ðŸ˜‰","type":"unstyled","depth":0,"inlineStyleRanges":[],"entityRanges":[],"data":{}}],"entityMap":{}}`,
//   errorMsg: '',
//   fans: ['CUbE9RoUIYVOzGp04UPVzQO68Xr1'],
//   name: 'Ma Superbe Vitrine',
//   phone: casual.phone,
//   photos: {
//     name: 'worpress.png',
//     path: 'stores/CUbE9RoUIYVOzGp04UPVzQO68Xr1/1558490933970_wordpress.png',
//     src:
//       'https://firebasestorage.googleapis.com/v0/b/project-zero-csprod.appspot.com/o/stores%2FCUbE9RoUIYVOzGp04UPVzQO68Xr1%2F1558490933970_wordpress.png?alt=media&token=02057971-89b9-4224-8f85-cb46428c7ff1'
//   },
//   siret: '522 948 256',
//   status,
//   tags: ['Restaurant', 'Wifi', 'Famille']
// })
