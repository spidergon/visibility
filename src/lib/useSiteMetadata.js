import { useStaticQuery, graphql } from 'gatsby'

const useSiteMetadata = () => {
  const { site } = useStaticQuery(
    graphql`
      query SiteMetaData {
        site {
          siteMetadata {
            title
            description
            author
            keywords
          }
        }
      }
    `
  )
  return site.siteMetadata
}

export default useSiteMetadata
