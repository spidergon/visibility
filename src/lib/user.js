import { useContext } from 'react'
import firebase, { auth } from './base'
import { UserContext } from '../components/UserProvider'

/**
 * Hook that gives the authentivated user
 * @returns {Object} { initializing, user }.
 */
export function useSession () {
  return useContext(UserContext)
}

/**
 * Sign In by a provider.
 * @param {string} provider - the provider.
 * @param {function} authHandler - the auth callback.
 * @param {function} setLoading - the loading callback.
 * @param {function} setError - the error callback.
 */
export function signIn (provider, authHandler, setLoading, setError) {
  auth
    .signInWithPopup(new firebase.auth[`${provider}AuthProvider`]())
    .then(authHandler)
    .catch(error => {
      setError({ code: error.code, message: error.message })
      setLoading(false)
    })
}

/** Sign In Anonymously. */
export function signInAnonymously () {
  return auth.signInAnonymously()
}

/**
 * Send e-mail containing a link to automatically sign in by e-mail.
 * @param {string} email - the email to send the link to.
 * @param {function} callback - the callback to trigger after email sent.
 */
export function sendSignInEmailLink (email, callback) {
  // TODO: change url (domain name)
  auth
    .sendSignInLinkToEmail(email, {
      url: 'http://localhost:8000/connexion?finishEmailSignIn',
      handleCodeInApp: true
    })
    .then(() => {
      window.localStorage.setItem('emailForSignIn', email)
      callback()
    })
    .catch(error => callback(error))
}

/**
 * Sign In by the link provided by e-mail.
 * @param {function} callback - the callback to trigger after signed in.
 */
export function signInEmailLink (callback) {
  if (auth.isSignInWithEmailLink(window.location.href)) {
    let email = window.localStorage.getItem('emailForSignIn')
    if (!email) {
      email = window.prompt('Veuillez confirmer votre adresse e-mail :') || ''
    }
    auth
      .signInWithEmailLink(email, window.location.href)
      .then(() => {
        window.localStorage.removeItem('emailForSignIn')
        callback()
      })
      .catch(error => callback(error))
  }
}

/** 'auth/account-exists-with-different-credential' error code. */
export const AUTH_CREDENTIAL_IN_USE =
  'auth/account-exists-with-different-credential'

/** Regex for validate email. */
export const EMAIL_REGEX = /^([\w.%+-]+)@([\w-]+\.)+([\w]{2,})$/i
