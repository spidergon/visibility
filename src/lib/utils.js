import md5 from 'md5'

/**
 * Loop over an array asynchronously.
 * @param {array} array - the array to loop over.
 * @param {function} callback - the callback to trigger for each item.
 */
export async function asyncLoop (array, callback) {
  for (let index = 0, len = array.length; index < len; index++) {
    await callback(array[index], index)
  }
}

// /**
//  * Read a file.
//  * @param {Object} file - the file object to read.
//  * @param {function} setError - the error callback.
//  * @returns {Promise} A promise containing the data.
//  */
// export function readFile (file, setError) {
//   const reader = new FileReader()
//   return new Promise((resolve, reject) => {
//     reader.onload = () => resolve(reader.result)
//     reader.onerror = () => {
//       reader.abort()
//       setError('Le chargement du fichier a échoué !')
//       reject(new DOMException('Problem parsing input file.'))
//     }
//     reader.readAsDataURL(file)
//   })
// }

/**
 * Returns the gravatar url for the given email
 * @param {string} email - The email to get the avatar from.
 */
export const gravatar = email =>
  `https://www.gravatar.com/avatar/${md5(email)}?d=mp`

/**
 * Slugify a name.
 * @param {string} name - the name to slugify.
 * @returns {string} the slugified name.
 */
export function slugify (name) {
  return require('slugify')(name, { remove: /[*+~.()'"!:@]/g, lower: true })
}

export const dashPath = '/app/dashboard'

// function geoloc (success) {
//   if ('geolocation' in navigator) {
//     navigator.geolocation.getCurrentPosition(
//       position => success(position),
//       error => {
//         switch (error.code) {
//           case error.PERMISSION_DENIED:
//             console.log('User denied the request for Geolocation.')
//             break
//           case error.POSITION_UNAVAILABLE:
//             console.log('Location information is unavailable.')
//             break
//           case error.TIMEOUT:
//             console.log('The request to get user location timed out.')
//             break
//           default:
//             console.log('Geolocation: an unknown error occurred.')
//         }
//       }
//     )
//   } else console.log('Geolocation is not supported by this browser.')
// }
