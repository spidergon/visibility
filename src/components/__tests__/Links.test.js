import React from 'react'
import { render } from '../../lib/test-utils'
import Link from '../Link'

describe('<Links/>', () => {
  it('renders "internal" links correctly', () => {
    const { container } = render(<Link to='/'>{'My Link'}</Link>)
    expect(container).toMatchSnapshot()
  })

  it('renders "internal file" links correctly', () => {
    const { container } = render(<Link to='/test.pdf'>{'My Link'}</Link>)
    expect(container).toMatchSnapshot()
  })

  it('renders "external" links correctly', () => {
    const { container } = render(
      <Link blank to='https://www.google.com'>
        {'My Link'}
      </Link>
    )
    expect(container).toMatchSnapshot()
  })
})
