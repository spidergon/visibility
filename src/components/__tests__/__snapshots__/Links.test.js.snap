// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Links/> renders "external" links correctly 1`] = `
<div>
  <a
    aria-label="https://www.google.com (sâ€™ouvre dans un nouvel onglet)"
    href="https://www.google.com"
    rel="noreferrer noopener nofollow"
    target="_blank"
  >
    My Link
  </a>
</div>
`;

exports[`<Links/> renders "internal file" links correctly 1`] = `
<div>
  <a
    href="/test.pdf"
  >
    My Link
  </a>
</div>
`;

exports[`<Links/> renders "internal" links correctly 1`] = `
<div>
  <a
    href="/"
  >
    My Link
  </a>
</div>
`;
